source/opt/unify_const_pass.cpp

```diff
@@ -0,0 +1,185 @@
+    auto keys = GetLookUpKeys(inst);
+    auto* node = root_.get();
+    for (uint32_t key : keys) {
+      node = node->GetOrCreateTrieNodeFor(key);
+    }
+    if (node->result_id() == 0) {
+      node->SetResultId(inst.result_id());
+    }
+    return node->result_id();
+  }
+
+ private:
+  // Extracts the opcode, result type id and operand words from an instruction.
+  // Returns a vector of those values in 32-bit unsigned integers.
+  std::vector<uint32_t> GetLookUpKeys(const ir::Instruction& inst) {
+    std::vector<uint32_t> keys;
+    // Need to use the opcode, otherwise there might be a conflict with the
+    // following case when <op>'s binary value equals xx's id:
+    //  OpSpecConstantOp tt <op> yy zz
+    //  OpSpecConstantComposite tt xx yy zz;
+    keys.push_back(static_cast<uint32_t>(inst.opcode()));
+    for (uint32_t i = 0; i < inst.NumOperands(); i++) {
+      if (inst.GetOperand(i).type == SPV_OPERAND_TYPE_RESULT_ID) continue;
+      keys.insert(keys.end(), inst.GetOperand(i).words.cbegin(),
+                  inst.GetOperand(i).words.cend());
+    }
+    return keys;
+  }
+
+  std::unique_ptr<ResultIdTrieNode> root_;  // The root node of the trie.
+};
+}  // anonymous namespace
+
+bool UnifyConstantPass::Process(ir::Module* module) {
+  bool modified = false;
+  ResultIdTrie defined_constants;
+  std::unordered_set<ir::Instruction*> duplicated_cosnts;
+  analysis::DefUseManager def_use_mgr(module);
+
+  for (ir::Instruction& inst : module->types_values()) {
+    // Do not handle the instruction when there are decorations upon either
+    // the result id or the type id.
+    if (def_use_mgr.GetAnnotations(inst.result_id()).size() != 0 ||
+        def_use_mgr.GetAnnotations(inst.type_id()).size() != 0) {
+      continue;
+    }
+
+    // The overall algorithm is to store the result ids of all the eligible
+    // constants encountered so far in a trie. For a comming constant defining
+    // instruction, use its opcode, result type id and operand words as an
+    // array of keys to lookup the trie. If an result id can be found for that
+    // array of keys, a constant with exactly the same value(s) must has been
+    // defined before, the constant under processing should be replace by the
+    // constant previously defined. If none result id can be found for that
+    // array of keys, this must be the first time a constant with its value
+    // be defined, we then create a new trie node to store the result id with
+    // the keys. When replacing a duplicated constant with a previously defined
+    // constant, all the uses of the duplicated constant, which must be placed
+    // after the duplicated constant defining instruction, will be updated.
+    // This way, the descendants of the previously defined constant and the
+    // duplicated constant will both refer to the previously defined constant.
+    // So that the operand ids that used in key arrays will always be ids of
```

Comment:
... will be ids of the unified constants, when processing is up to a descendant.